### ISSUE TITLE
Promise unwrap fails as a whole even if some of the requests can still be fulfilled

### ISSUE THREAD

**Issue by bizmate** [guzzle/guzzle#1481]
I am trying to run multiple requests at the same time against totally separate domains. When one domain fails the whole promise returns an Exception. In my view this is not flexible enough and cannot be controlled.

Here is an example, all in a symfony controller for simplicity

```
public function guzzlePromiseAction()
    {
        $logger = $this->get('logger');
        $siteArr = [
            '00skka5ae8ofkskar3umc1f11tbrni76.com',
            '00tamago.com',
            'google.com'
        ];
        $sitesChecked = [];

        $client = new Client();
        foreach($siteArr as $site){
            $requestArr[$site] = $client->getAsync( 'http://' . $site );
        }

        try{
            $time_sentiment_promise_start = microtime(true);
            $responses = Promise\unwrap($requestArr);
            $time_sentiment_promise_end = microtime(true);


            $time_sentiment_bodies_start = microtime(true);

            foreach($responses as $domain => $response){
                $hasBody = strlen( $response->getBody() );
                $sitesChecked[$domain] =  $hasBody;
            }

            $time_sentiment_bodies_end = microtime(true);

            $logger->info(' Crawler TEST: check took ' .
                ($time_sentiment_bodies_end - $time_sentiment_bodies_start) .
                ' promise execution ' .
                ($time_sentiment_promise_end - $time_sentiment_promise_start)
            );
        }
        catch(\Exception $e){
            $logger->info(' Crawler TEST:  exception line: ' .$e->getLine() .
                " msg:" . $e->getMessage() . ' --- Trace:' . $e->getTraceAsString()
            );
        }
        exit;
    }
```

The exception logged is 

```
root@a80f91e6777c:/var/www/symfony/app/logs# tail -f dev.log | grep "Crawler TEST"
[2016-05-23 07:50:56] app.INFO:  **Crawler TEST:  exception line: 186 msg:cURL error 6: Could not resolve host: 00skka5ae8ofkskar3umc1f11tbrni76.com** (see http://curl.haxx.se/libcurl/c/libcurl-errors.html) --- Trace:#0 /var/www/symfony/vendor/guzzlehttp/guzzle/src/Handler/CurlFactory.php(150): GuzzleHttp\Handler\CurlFactory::createRejection(Object(GuzzleHttp\Handler\EasyHandle), Array) #1 /var/www/symfony/vendor/guzzlehttp/guzzle/src/Handler/CurlFactory.php(103): GuzzleHttp\Handler\CurlFactory::finishError(Object(GuzzleHttp\Handler\CurlMultiHandler), Object(GuzzleHttp\Handler\EasyHandle), Object(GuzzleHttp\Handler\CurlFactory)) #2 /var/www/symfony/vendor/guzzlehttp/guzzle/src/Handler/CurlMultiHandler.php(179): GuzzleHttp\Handler\CurlFactory::finish(Object(GuzzleHttp\Handler\CurlMultiHandler), Object(GuzzleHttp\Handler\EasyHandle), Object(GuzzleHttp\Handler\CurlFactory)) #3 /var/www/symfony/vendor/guzzlehttp/guzzle/src/Handler/CurlMultiHandler.php(108): GuzzleHttp\Handler\CurlMultiHandler->processMessages() #4 /var/www/symfony/vendor/guzzlehttp/guzzle/src/Handler/CurlMultiHandler.php(123): GuzzleHttp\Handler\CurlMultiHandler->tick() #5 /var/www/symfony/vendor/guzzlehttp/promises/src/Promise.php(242): GuzzleHttp\Handler\CurlMultiHandler->execute(true) #6 /var/www/symfony/vendor/guzzlehttp/promises/src/Promise.php(219): GuzzleHttp\Promise\Promise->invokeWaitFn() #7 /var/www/symfony/vendor/guzzlehttp/promises/src/Promise.php(262): GuzzleHttp\Promise\Promise->waitIfPending() #8 /var/www/symfony/vendor/guzzlehttp/promises/src/Promise.php(221): GuzzleHttp\Promise\Promise->invokeWaitList() #9 /var/www/symfony/vendor/guzzlehttp/promises/src/Promise.php(62): GuzzleHttp\Promise\Promise->waitIfPending() #10 /var/www/symfony/vendor/guzzlehttp/promises/src/functions.php(192): GuzzleHttp\Promise\Promise->wait() #11 /var/www/symfony/src/AppBundle/Controller/DevController.php(160): GuzzleHttp\Promise\unwrap(Array) #12 [internal function]: AppBundle\Controller\DevController->guzzlePromiseAction() #13 /var/www/symfony/app/bootstrap.php.cache(3210): call_user_func_array(Array, Array) #14 /var/www/symfony/app/bootstrap.php.cache(3172): Symfony\Component\HttpKernel\HttpKernel->handleRaw(Object(Symfony\Component\HttpFoundation\Request), 1) #15 /var/www/symfony/app/bootstrap.php.cache(3323): Symfony\Component\HttpKernel\HttpKernel->handle(Object(Symfony\Component\HttpFoundation\Request), 1, true) #16 /var/www/symfony/app/bootstrap.php.cache(2514): Symfony\Component\HttpKernel\DependencyInjection\ContainerAwareHttpKernel->handle(Object(Symfony\Component\HttpFoundation\Request), 1, true) #17 /var/www/symfony/web/app_dev.php(28): Symfony\Component\HttpKernel\Kernel->handle(Object(Symfony\Component\HttpFoundation\Request)) #18 {main} [] []
```

I would assume you could handle an exception for each of the promise return types instead of having the whole set of multiple requests fail. Any thoughts?

**Comment by Tobion** [comment]
What was your conclusion considering you closed it?

**Comment by bizmate** [comment]
I think documentation was updated from last time i used Promises. Now if you look at the concurrent requests section in 
http://docs.guzzlephp.org/en/latest/quickstart.html

```
// Wait on all of the requests to complete. Throws a ConnectException
// if any of the requests fail
$results = Promise\unwrap($promises);

// Wait for the requests to complete, even if some of them fail
$results = Promise\settle($promises)->wait();
```

My approach was throwing a connection exception as i was using unwrap. Settle and wait works better. I also tried the Pool approach but didnt like it as much. Now i am able to run about 100 requests within 15-20sec forcing a timeout.
